Enable autoformatting

bitwiseman opened the pull request:
# Description
Formatting is a continuing issue during code reviews.  Different users' IDE's have different opinions on what is correct formatting and those also differ from when various reviewers consider correct. 

This change enables a base set of formatting and import ordering rules, automating the process of formatting code and enforcing that formatting. The setting I've used are somewhat arbitrary, but at least attempt to not completely torpedo the existing formatting. 

This change uses the `set.changelist` property (set only in CI) to switch from applying formatting to checking formatting. On developer machines, the change will automatically apply formatting.  In CI, formatting will be checked not updated.

Once this PR is merged, I will go back and merge existing PRs with https://github.com/bitwiseman/github-branch-source-plugin/tree/task/formatting-base, rebuild them, and then merge in master.  This should result in few to no merge conflicts. 

I am aware that this is controversial, so I've included a range of people as reviewers and am happy to have more weigh in. However, I need to be clear that I consider the lack of this to be technical organizational debt that  adds one more hurdle to entry by new contributors and adds time to every PR review.

If this works here, I plan to apply it to other plugins, at least those I am a maintainer on.

bitwiseman commented on src/main/java/org/jenkinsci/plugins/github_branch_source/ApiRateLimitChecker.java:
```diff
@@ -42,46 +42,48 @@ long checkRateLimitImpl(@NonNull GHRateLimit.Record rateLimit, long count, long
 
-    /**å
-     * Restrict GitHub API requests only when near or above rate limit.
+    /**
+     * å Restrict GitHub API requests only when near or above rate limit.
```
```suggestion
    /**
     * Restrict GitHub API requests only when near or above rate limit.
```

timja reviewed (approved):
If it’s automated I really don’t care much about the specifics, it shouldn’t be controversial imo, much better to let automation handle it than have humans worrying about it

car-roll commented:
+1
definitely prefer this over an IDE accidentally autoformatting an entire class that gets pushed up. Not to mention feeling like some sort of style gatekeeper when looking over a review and seeing something that doesn't quite adhere to the coding convention of the particular plugin.

alecharp commented:
:100: I feel like any automation to normalize the code in a PR to the rest of the project code is a plus. This way we can really focus on what the code is doing rather than being distracted by how the code is written.

rsandell commented on src/test/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigatorTraitsTest.java:
```diff
@@ -364,1200 +273,775 @@ public void limit_branches() throws Exception {
-                        )
-                )
-        );
+                        Matchers.allOf(instanceOf(OriginPullRequestDiscoveryTrait.class),
```
Here the auto formatting has made it less readable than before.

MRamonLeon replied:
Agree, we should get used to it, I guess

rsandell commented on src/test/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigatorTraitsTest.java:
```diff
@@ -364,1200 +273,775 @@ public void limit_branches() throws Exception {
+                        Matchers.allOf(instanceOf(OriginPullRequestDiscoveryTrait.class),
+                                hasProperty("strategyId", is(2))),
+                        Matchers.allOf(instanceOf(ForkPullRequestDiscoveryTrait.class),
+                                hasProperty("strategyId", is(2)))));
```
Here it's very bad imho :)

jglick commented on src/main/java/org/jenkinsci/plugins/github_branch_source/ApiRateLimitChecker.java:
```diff
@@ -4,17 +4,16 @@
+import org.jenkinsci.plugins.github.config.GitHubServerConfig;
+import org.kohsuke.github.GHRateLimit;
+import org.kohsuke.github.GitHub;
+import org.kohsuke.github.RateLimitChecker;
```
Switched to a simple alphabetical order IIUC? Very welcome. Humans should not be reading `import` statements to begin with; normalizing the order minimizes the chance of stupid merge conflicts.

bitwiseman replied:
@jglick  
It is my understanding that what code style guidelines there are for Jenkins say "Simple alphabetical non-static, then simple alphabetical static".   Is that incorrect? 

@jtnord 
What you prefer the ordering be?  I don't care what it is as long as I never have to tell someone in a code review to change it.

jglick replied:
So long as it is simple I would not really care. (Whether `static` `import`s should be allowed at all, and under what conditions, is a contentious topic.)

bitwiseman replied:
@jglick 
The topic of `static` `import`s will not be solved by a formatter.

jtnord replied:
>What you prefer the ordering be? I don't care what it is as long as I never have to tell someone in a code review to change it.

I have personal preferences but I do not care too much about the order apart from always non static before static imports though, and static imports should be reserved for test code apart from those very rare special occasions.

But given as you asked, my preference is someone along the following lines.

1. core libraries first (`java`, `javax`)
2. then 3rd party libraries (`com` .`edu` , `org`)
3. then a break 
4. then jenkins core code (`hudson`, `jenkins`, `org.jenkinsci`, `io.jenkins`)
5. then stuff from my own project (e.g. `org.jenkinsci.plugins.github_branch_source`)
6. then a break
7. then the same but for static imports.

That said my IDE will automatically an import put it where it is configured to by its ordering (eclipse) and having per project settings is a PITA (I do not recall the last time I typed an import statement).  I try to reduce diff bloat it after the fact by manually moving it if I recall :)


> The topic of static imports will not be solved by a formatter.

you mean the ordering or that they should or should not be allowed?

bitwiseman replied:
Whether or not they should be allowed.

jglick replied:
FWIW my personal preference is strictly alphabetical, no line breaks, no attempt to group in any way other than `| sort`.

jglick commented on src/main/java/org/jenkinsci/plugins/github_branch_source/ApiRateLimitChecker.java:
```diff
@@ -42,46 +42,48 @@ long checkRateLimitImpl(@NonNull GHRateLimit.Record rateLimit, long count, long
                     if (expiration != now) {
-                        writeLog("Jenkins is attempting to evenly distribute GitHub API requests. To configure a different rate limiting strategy, such as having Jenkins restrict GitHub API requests only when near or above the GitHub rate limit, go to \"GitHub API usage\" under \"Configure System\" in the Jenkins settings.");
+                        writeLog(
+                                "Jenkins is attempting to evenly distribute GitHub API requests. To configure a different rate limiting strategy, such as having Jenkins restrict GitHub API requests only when near or above the GitHub rate limit, go to \"GitHub API usage\" under \"Configure System\" in the Jenkins settings.");
```
Oof, this looks pretty bad. Maybe rules pertaining to line length (which I guess is responsible for this) can be suppressed?

bitwiseman replied:
@jglick @jtnord 
I tried changing the method name to `log` and got:
```
                    if (expiration != now) {
                        log("Jenkins is attempting to evenly distribute GitHub API requests. To configure a different rate limiting strategy, such as having Jenkins restrict GitHub API requests only when near or above the GitHub rate limit, go to \"GitHub API usage\" under \"Configure System\" in the Jenkins settings.");
                    }
```

:shrug:  
Apparently, even one character improvement causes wrapping.  I take this as another case where the formatter is highlighting not great code by making it uglier.

If there were an option to `never wrap first argument`, I'd use it.  But there isn't so, I'll take this.

jtnord replied:
> If there were an option to never wrap first argument,

comes back to auto formatters are not as smart as a good person.

bitwiseman replied:
> > If there were an option to never wrap first argument,

> comes back to auto formatters are not as smart as a good person.

And [John Henry beat the steam engine](https://en.wikipedia.org/wiki/John_Henry_(folklore)), but I don't want to be John Henry.  😄

bitwiseman replied:
This can be addressed by making the method name shorter or by manually splitting the string onto multiple lines.  I've opted for the latter.

jglick commented on src/main/java/org/jenkinsci/plugins/github_branch_source/BranchDiscoveryTrait.java:
```diff
@@ -77,7 +77,8 @@
      *
-     * @param strategyId the strategy id.
+     * @param strategyId
+     *            the strategy id.
```
This is less legible.

bitwiseman replied:
Okay, fixed.

jglick commented on src/main/java/org/jenkinsci/plugins/github_branch_source/Endpoint.java:
```diff
@@ -59,13 +57,7 @@
-            "scm.",
-            "source."
-    };
+    private static final String[] COMMON_PREFIX_HOSTNAMES = { "git.", "github.", "vcs.", "scm.", "source." };
```
Preferred the previous form, which makes it easier to review diffs to the list.

bitwiseman replied:
@jglick 
Agreed, but :shrug: .
Do we want all array initializers to have one line per item?  Or is this a matter of preference for this specific case?

jglick replied:
Initializers which might reasonably be edited in the future.

bitwiseman replied:
Addressed by forcing expansion using a line comment.  A little hacky but lets us keep most arrays unexpanded, which seems to be the more generally expected behavior.

jglick commented on src/main/java/org/jenkinsci/plugins/github_branch_source/Endpoint.java:
```diff
@@ -151,15 +143,17 @@ public FormValidation doCheckApiUri(@QueryParameter String apiUri) {
                 LOGGER.log(Level.FINE, "Trying to configure a GitHub Enterprise server");
-                // For example: https://api.github.com/ or https://github.mycompany.com/api/v3/ (with private mode disabled).
+                // For example: https://api.github.com/ or https://github.mycompany.com/api/v3/ (with private mode
+                // disabled).
```
Not helpful IMO.

bitwiseman replied:
:shrug: 
Okay, this can be controlled. Line comments to do not join if already wrapped, so user can wrap lines earlier if they choose.

jglick commented on src/main/java/org/jenkinsci/plugins/github_branch_source/ForkPullRequestDiscoveryTrait.java:
```diff
@@ -392,7 +402,9 @@ public boolean isApplicableToOrigin(@NonNull Class<? extends SCMHeadOrigin> orig
-    public static class TrustEveryone extends SCMHeadAuthority<SCMSourceRequest, PullRequestSCMHead, PullRequestSCMRevision> {
+    public static class TrustEveryone
+            extends
+                SCMHeadAuthority<SCMSourceRequest, PullRequestSCMHead, PullRequestSCMRevision> {
```
:sob:

bitwiseman replied:
Fixed.

jglick commented on src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubBuildStatusNotification.java:
```diff
@@ -55,13 +55,10 @@
- * Job (associated to a PR) scheduled: PENDING
- * Build doing a checkout: PENDING
- * Build done: SUCCESS, FAILURE or ERROR
+ * Job (associated to a PR) scheduled: PENDING Build doing a checkout: PENDING Build done: SUCCESS, FAILURE or ERROR
```
This is mangled.

jtnord replied:
should really have been in a definition list or the like (eclipse here is at least happy about not reformatting that)

bitwiseman replied:
Yeah, this highlights false assumption on the part of the person writing the javadoc.

jglick commented on src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubNotificationRequest.java:
```diff
@@ -28,12 +28,12 @@
+ * Context label. If two notification requests with the same Context label are provided, one will override the other.
+ *
+ * @see <a href="https://developer.github.com/v3/repos/statuses/">Github API</a> for details of the purpose of each
+ *      notification field.
```
This appears to have broken Javadoc (see check below).

jglick commented on src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubNotificationRequest.java:
```diff
@@ -28,12 +28,12 @@
- * Notifications are differentiated by their Context label. If two notification requests with the same Context label are
- * provided, one will override the other.
- * @see <a href="https://developer.github.com/v3/repos/statuses/">Github API</a> for details of the purpose of each notification field.
+ * Details of a GitHub status notification to be sent. One GitHubNotificationRequest represents one notification. A
```
The summary sentence is distinguished in Javadoc and ought to be on its own source code line.

jglick commented on src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubRepositoryInfo.java:
```diff
@@ -25,19 +25,18 @@
- * Is it worth making a guess based on the specified host?
+ *
+ * TODO: Is it possible to compute the API URI from just a repository URL, or not because of the possibility of proxies,
+ * etc.? Is it worth making a guess based on the specified host?
```
Breaks at an illogical point.

bitwiseman replied:
:shrug: 
It considers this text, it wraps the text.

jglick reviewed (commented):
Helpful in some spots but personally I find this a net negative in terms of legibility. Mechanical formatting is just terrible at deciding where to insert line breaks.

Any change like this implies a “diff wall”: for the promise of reduced merge conflicts, review overhead, etc. going forward, you pay the cost of `git blame` becoming much less useful for understanding older code, and cherry-picking changes to backport branches preceding the cutoff becomes much harder.

jtnord commented on src/build/eclipse/eclipse.importorder:
```diff
@@ -0,0 +1,9 @@
+# 1=java
+# 2=javax
+#
+# static imports
```
this is all commented out - what does it acheive?

bitwiseman replied:
Kept it to show the alternative settings, fine with removing.

jtnord commented on src/build/eclipse/formatter.xml:
```diff
@@ -0,0 +1,387 @@
+        <setting id="org.eclipse.jdt.core.formatter.insert_space_before_colon_in_default" value="do not insert"/>
+        <setting id="org.eclipse.jdt.core.formatter.tabulation.char" value="space"/>
+    </profile>
+    <!-- Keeping the "default" exported values for reference -->
```
the eclipse default is terrible....

bitwiseman replied:
🤷  okay.

jtnord commented on src/main/java/org/jenkinsci/plugins/github_branch_source/ApiRateLimitChecker.java:
```diff
@@ -96,12 +98,14 @@ public LocalChecker getChecker(@NonNull TaskListener listener, String apiUrl) {
                 // If the GitHub public API is being used, this will fallback to ThrottleOnOver
                 LocalChecker checker = ThrottleOnOver.getChecker(listener, apiUrl);
-                checker.writeLog("GitHub throttling is disabled, which is not allowed for public GitHub usage, so ThrottleOnOver will be used instead. To configure a different rate limiting strategy, go to \"GitHub API usage\" under \"Configure System\" in the Jenkins settings.");
+                checker.writeLog(
```
should be aligned to `writeLog(` not `checker`.

basically are you using a fixed offset for wrapped indentation?  (ie 8 which is the lengh of `checker.` - in which case don't  use an offset that matches to the thing you are offsetting.

otherwise you can see something like

```
checkz(.writeLog(
            "a very long string),
            "and something else")
```

and that is worse than evil.

bitwiseman replied:
@jtnord 
Same debatable point as above.   The indent you are saying is missing was (as far as I can tell) intentionally removed as redundant.

bitwiseman replied:
@jtnord 
Regarding the align to column, you can change this the formatter as noted  [here](https://github.com/jenkinsci/github-branch-source-plugin/pull/392/files#r580691521) to be `82` and it will align to column, but it produces results that I think you'll like even less than the current results.  `50` is also an option, but it ends up even worse. 

But maybe I'm misunderstanding what you mean.

jtnord replied:
so given the settings that are being used are offset by a specific amount (and not column) if it hard to see when something is wrapped in something like

```
foo(bar(wibble(baz, bob,
        jennygoestomarket),
        markstaysathome));
```

Because both `jennygoestomarket` and `markstaysathome` are `offset` by the same amount, and when scanning the code it looks like they can be the arguments to the same function (when they are not).
Whitespace is helpful to show what things apply to (which is why we indent blocks `{`)  to me the same applies here.
Maybe I misunderstand the settings you are applying? (but I have had many issues with eclipse when using this setting in the past)

disclaimer \#1  - I much prefer projects that are set to indent on column.
disclaimer \#2 I am mildly dyslexic and so this is important to me.  **My** brain scans and reads text in a specific and different to the vast majority of people.

now - I am not a major contributor to this project and so I am not trying to enforce my standards onto the people who do work with it much more than I do (I have only had a few minor contributions) and as such I will adapt when I need to work in this project.  However if this is the first step into trying to get something more standard into Jenkins ecosystem and rolling out more widely then being inclusive of all people (which includes the neuro diverse) should be taken into account, as descisions made here can make it more or less easy for people to contribute.

disclaimer \#3  on my projects when I remember to switch to *my* coding style I `indent` with tabs so that people can customize their indent view to support their needs (2 for some people 8 for others, and some have 4 or even 3) but `offset` with spaces.  this allows the code to look the same regardless of tab space setting and allow people to have a visual offset that conforms to the way they work best.

and finally disclaimer \#4 all attempts at auto formatting will break somewhere.  This is why I think auto formatters are evil.  let the bot do a formatting review, let a human decide to reject it's findings, or to easily make it better.  See the end part on disclaimer 2

bitwiseman replied:
@jtnord 
1. The problem I've seen with with indenting to column is you end up with deeply indented code.   I assume your suggestion for addressing that is to refactor the code into smaller methods, avoiding deeply nested structures. Am I correct? 

2. Neuro-atypicalness and accessibility is certainly a valid concern.  I read slowly and also depend on formatting to allow me to take in code at a reasonable speed.  I definitely want to be sensitive to a broad range of contributors needs.  At the same time, consider that someone else might find  "align to column" harder to read due to its deeper indenting causing lines to wrap more often.  Which one wins? In the case of this change, I am looking to apply a standard formatting that most closely matches what is already being used. There are parts of this code base that seem to be using "align to column", but the majority of it does not.  

3. I believe this mix of tabs and spaces is referred to as "smart-tabs". It works best in "align to column" if I understand it correctly.  I'm not opposed to this, but as with point two it seems a larger change.   I've done like I did with [no-join](https://github.com/jenkinsci/github-branch-source-plugin/compare/master...bitwiseman:task/formatting-no-join?expand=1) and created a branch using this and "align to column" settings:  [bitwiseman:task:task/formatting...bitwiseman:task/formatting-columns](https://github.com/bitwiseman/github-branch-source-plugin/compare/task/formatting...bitwiseman:task/formatting-columns?expand=1).  Perhaps there's other settings that could be applied to do this better.  Perhaps you as the one bringing this up could try various setting to find one that is best in your opinion? 

4. I would instead say that **all attempts at formatting will be imperfect**, both humans and automated systems.  Auto-formatters are generally imperfect in rigid and predictable ways, whereas human formatters are imperfect in a wide range of random ways.  I categorically disagree that humans should be allowed to opt-out of auto-formatting - given the option humans will choose to do things their own way and then we end up talking about "better" or "worse" formatting instead of focusing on making the code better.

jglick replied:
My preference would be to use the autoformatter for things it can do reliably—sort imports, remove trailing whitespace, enforce spaces vs. tabs for block indentation, normalize use of spaces around / inside tokens like `(` `)` `=`—but for more subjective things like placement of line breaks leave the choice to humans.

jtnord commented on src/main/java/org/jenkinsci/plugins/github_branch_source/ForkPullRequestDiscoveryTrait.java:
```diff
@@ -85,26 +85,34 @@
     @DataBoundConstructor
-    public ForkPullRequestDiscoveryTrait(int strategyId,
-                                         @NonNull SCMHeadAuthority<? super GitHubSCMSourceRequest, ? extends ChangeRequestSCMHead2, ? extends SCMRevision> trust) {
+    public ForkPullRequestDiscoveryTrait(
+            int strategyId,
+            @NonNull SCMHeadAuthority<? super GitHubSCMSourceRequest, ? extends ChangeRequestSCMHead2, ? extends SCMRevision> trust) {
```
worse than before by a long way.
see previous comment about indention based on a fixed amount vs relative.

bitwiseman replied:
@jtnord 
Debatable. Do you really want to debate this?

jtnord commented on src/main/java/org/jenkinsci/plugins/github_branch_source/ForkPullRequestDiscoveryTrait.java:
```diff
@@ -85,26 +85,34 @@
+    public ForkPullRequestDiscoveryTrait(
+            @NonNull Set<ChangeRequestCheckoutStrategy> strategies,
+            @NonNull SCMHeadAuthority<? super GitHubSCMSourceRequest, ? extends ChangeRequestSCMHead2, ? extends SCMRevision> trust) {
+        this(
+                (strategies.contains(ChangeRequestCheckoutStrategy.MERGE) ? MERGE : NONE)
+                        + (strategies.contains(ChangeRequestCheckoutStrategy.HEAD) ? HEAD : NONE),
+                trust);
```
worse than before IMO

bitwiseman replied:
@jtnord 
Agreed, but I'm not sure the code itself was great to begin with.

jtnord reviewed (commented):
Personally I have never met an autoformatter that works in all cases.  so this leads me to ask 
what are the equivalent `//@autoformat:off` `//@autoformat:on` flags and will it make the code more or less useable.

I also dislike things that change code just because I build, I like to know what happens - sure let the build fail if I need to run `mvn -Pfix-the-formatting validate` then let it be so.

jglick commented:
> I also dislike things that change code just because I build, I like to know what happens - sure let the build fail if I need to run `mvn -Pfix-the-formatting validate` then let it be so.

:100:, principle of least surprise says that commands like `mvn clean verify` should never modify my working tree (or leave untracked files).

I feel like there was some detailed proposal for formatting enforcement a year or two ago which acknowledged this, but not sure where that was. Cannot find a JEP on the topic. Maybe some ML thread?

jglick commented:
Ah there it is: https://groups.google.com/g/jenkinsci-dev/c/zUewd9LL5-g/m/t1AsbKOACQAJ

bitwiseman commented on src/main/java/org/jenkinsci/plugins/github_branch_source/ApiRateLimitChecker.java:
```diff
@@ -42,46 +42,48 @@ long checkRateLimitImpl(@NonNull GHRateLimit.Record rateLimit, long count, long
-                                        Util.getTimeSpanString(rateLimitResetMillis),
-                                        // The GitHubRateLimitChecker adds a one second sleep to each notification loop
-                                        Util.getTimeSpanString(1000 + expiration - now)));
+                                    "Jenkins-Imposed API Limiter: Current quota for Github API usage has %d remaining (%d over budget). Next quota of %d in %s. Sleeping for %s.",
+                                    rateLimit.getRemaining(),
+                                    ideal - rateLimit.getRemaining(),
+                                    rateLimit.getLimit(),
+                                    Util.getTimeSpanString(rateLimitResetMillis),
+                                    // The GitHubRateLimitChecker adds a one second sleep to each notification loop
+                                    Util.getTimeSpanString(1000 + expiration - now)));
```
That is debatable. There is only one argument to `writeLog` meaning that indent is redundant and can be skipped in this case.  If `writelLog` had another argument I would expect the children of `format` to be indented.  

Tested and got the expected output: 

```
                            writeLog("FINE",
                                    String.format(
                                            "Jenkins-Imposed API Limiter: Current quota for Github API usage has %d remaining (%d over budget). Next quota of %d in %s. Sleeping for %s.",
                                            rateLimit.getRemaining(),
                                            ideal - rateLimit.getRemaining(),
                                            rateLimit.getLimit(),
                                            Util.getTimeSpanString(rateLimitResetMillis),
                                            // The GitHubRateLimitChecker adds a one second sleep to each notification
                                            // loop
                                            Util.getTimeSpanString(1000 + expiration - now)));
```

jtnord commented:
I've come to the conclusion that standards are best enforced as review comments from a bot in a PR.

This way the code owner can disregard it where it is less readable, and as it comes from a bot it is not personal.

Provide me with an easy way to know how to format the code (which could be an explicit maven goal or a profile/phase)

I think the config-as-code plugin does that (may be some other repo I am mixing it up with)

bitwiseman commented on src/build/eclipse/formatter.xml:
```diff
@@ -0,0 +1,368 @@
+        <setting id="org.eclipse.jdt.core.formatter.alignment_for_arguments_in_annotation" value="48"/>
+        <setting id="org.eclipse.jdt.core.formatter.alignment_for_arguments_in_enum_constant" value="48"/>
+        <setting id="org.eclipse.jdt.core.formatter.alignment_for_arguments_in_explicit_constructor_call" value="48"/>
+        <setting id="org.eclipse.jdt.core.formatter.alignment_for_arguments_in_method_invocation" value="80"/>
```
@jtnord 
Changing this setting to `82` will align to the column.  I'll reference this in discussion of relative vs fixed.

bitwiseman commented on src/main/java/org/jenkinsci/plugins/github_branch_source/ApiRateLimitChecker.java:
```diff
@@ -26,7 +25,8 @@
             return new LocalChecker(listener) {
                 @Override
-                long checkRateLimitImpl(@NonNull GHRateLimit.Record rateLimit, long count, long now) throws InterruptedException {
+                long checkRateLimitImpl(@NonNull GHRateLimit.Record rateLimit, long count, long now)
+                        throws InterruptedException {
```
I can see the argument that `throws` should go on the same line.  But whether this less readable is debatable - do you really want to debate it?

jtnord replied:
you can choose to disregard my opinion here and I am ok with that.
hence 
> 🤷 you choose something and it gives you something.

auto formatters will always produce something that is worse.  this is to me one (all be it a very minor) case of that.

bitwiseman commented:
@rsandell 
Out of curiosity regarding the less than ideal formatting you pointed out, I tried adding "join wrapped lines = false" to the settings.  It addressed this issue and also some of the objections from @jglick (array initialization on multiple lines no longer get collapsed).  However, the result is significantly less consistent and more prone to odd hanging clauses - the user added a new line somewhere at some point and it just stays there.   

You can see the result here:
https://github.com/jenkinsci/github-branch-source-plugin/compare/task/formatting-no-join...bitwiseman:task/formatting-no-join?expand=1

bitwiseman commented:
@jglick @jtnord @rsandell 
I really appreciate your feedback on things that looked bad.  I have updated the setting to address that feedback where possible.   

There's a ton to unpack here.  And the volume of topics to unpack is where PR's like this often die.  There's a sufficient volume of objections and points of contention that the change does not get approved.  

The counter example is [`github-api`](https://github.com/hub4j/github-api).  I enabled this system on github-api and it has resulted in my never having to tell anyone to fix their formatting.  I don't even have to consider if I need to suggest a different formatting because they might have chosen to ignore the suggested formatting. This has allowed everyone to focus on implementing features and fixes resulting in more contributions from a broader range of people.  Jesse did indeed find some cases where the formatting was not great, but I think that is indicative of the API being used being rather clunky as well. 


> I've come to the conclusion that standards are best enforced as review comments from a bot in a PR.
> 
> This way the code owner can disregard it where it is less readable, and as it comes from a bot it is not personal.

And I've come to the conclusion that my time is my most scarce resource, and I don't want to spend any time dealing with different contributors opinions of what is "less readable".  




> > I also dislike things that change code just because I build, I like to know what happens - sure let the build fail if I need to run `mvn -Pfix-the-formatting validate` then let it be so.
> 
> 💯, principle of least surprise says that commands like `mvn clean verify` should never modify my working tree (or leave untracked files).

From a build correctness perspective, we should not modify files during the build. I've changed the behavior to check by default and the user can manually run apply when they are ready.



> Any change like this implies a “diff wall”: for the promise of reduced merge conflicts, review overhead, etc. going forward, you pay the cost of `git blame` becoming much less useful for understanding older code, and cherry-picking changes to backport branches preceding the cutoff becomes much harder.

Yes, the "diff wall" is real and is a cost.  We will pay the cost now  and be done with.

bitwiseman commented:
From the email list: 

I've updated https://github.com/jenkinsci/github-branch-source-plugin/pull/392 with the following changes: 
* Source will not be changed during the build unless the user runs a specific maven goal
** "spotless:check" runs near the end of the lifecycle during the verify phase during local build.  Formatting issues will fail the build, source remains unchanged. 
** "spotless:check" runs near the start  of the lifecycle during the process-sources phase during CI. Formatting issues will fail the build, source remains unchanged. 
** Users may run "mvn spotless:apply" to automatically fix formatting.  This command is easy and short.  When "spotless:check" fails it specifically tells users to run "mvn spotless:apply" to fix formatting.
* I have adjusted the formatting rules to address some of the points raised during review
** Line comments do not join if wrapped - if you have comments that you want to not be reformatted, keep them under 120 chars and use line comments
** Javadoc formatting made a bit more compact - https://github.com/jenkinsci/github-branch-source-plugin/pull/392/commits/ea6e28ca69ea24a59584fba3c6190b145dc0925e
** Found a way to force array initializer expansion if that is what is desired - https://github.com/jenkinsci/github-branch-source-plugin/pull/392/commits/43a99508ab640db724e8b673c61855cb57670b82
** Long string arguments getting pushed to the next line?  Maybe the humans involved should split those up. https://github.com/jenkinsci/github-branch-source-plugin/pull/392#discussion_r580403349
** I've applied the ".gitattributes" suggested by Joseph.  The build now passes on Windows.
* I have also not addressed a number of other places where people have said the formatting looks "worse". These cases are generally debatable/personal-preference.  In some of these cases I agree with the assessment but I don't see it as a blocker.  If anyone disagrees with that evaluation, feel free to say so and explain why.  
** https://github.com/jenkinsci/github-branch-source-plugin/pull/392#discussion_r580104724 - see https://github.com/bitwiseman/github-branch-source-plugin/pull/new/task/formatting-test-cleanup
** https://github.com/jenkinsci/github-branch-source-plugin/pull/392#discussion_r580105388 - same as previous
** https://github.com/jenkinsci/github-branch-source-plugin/pull/392#discussion_r580402264
** https://github.com/jenkinsci/github-branch-source-plugin/pull/392#discussion_r580510020
** https://github.com/jenkinsci/github-branch-source-plugin/pull/392#discussion_r580513803

Note: I am using the eclipse formatter rather than to google formatter in an attempt to reduce the severity of the diff wall and disruption to folks familiar with the current code style. There are still significant differences, but they are far less than switching to google formatting.  However, that is still an option. You can see what it would look like here: https://github.com/bitwiseman/github-branch-source-plugin/pull/new/task/formatting-google 

@basil has made a strong case for switching to Google Formatter, which would obviates my tweaks above. If I follow my own stated reasoning to its logical conclusion, what's he's saying makes sense.  On the other hand, the "diff wall" in that case is  absolute. I need to sit with this for a day or two, but I will likely make that change.

jtnord commented:
Thanks for the explanations.

Based on those I believe the approach that Liam is taking is the best one
(that there is no auto formatting done, but provide a way to format it and
check the formatting in the CI).  I would still like to see line comments
in the CI rather than just an overall failure and error log, but I guess
given the limitations of what we have it is a good compromise.

as for the `diff-wall`  I have no issue with that.  git blame and
everything still works, you just may have to look at the parent of the diff
that changed the line - but that is still the case today - e.g.
refactoring.  I know others have a different opinion, just voicing mine.

/James

On Fri, 26 Feb 2021 at 03:12, Liam Newman <notifications@github.com> wrote:

> From the email list:
>
> I've updated #392
> <https://github.com/jenkinsci/github-branch-source-plugin/pull/392> with
> the following changes:
>
>    - Source will not be changed during the build unless the user runs a
>    specific maven goal
>    ** "spotless:check" runs near the end of the lifecycle during the
>    verify phase during local build. Formatting issues will fail the build,
>    source remains unchanged.
>    ** "spotless:check" runs near the start of the lifecycle during the
>    process-sources phase during CI. Formatting issues will fail the build,
>    source remains unchanged.
>    ** Users may run "mvn spotless:apply" to automatically fix formatting.
>    This command is easy and short. When "spotless:check" fails it specifically
>    tells users to run "mvn spotless:apply" to fix formatting.
>    - I have adjusted the formatting rules to address some of the points
>    raised during review
>    ** Line comments do not join if wrapped - if you have comments that
>    you want to not be reformatted, keep them under 120 chars and use line
>    comments
>    ** Javadoc formatting made a bit more compact - ea6e28c
>    <https://github.com/jenkinsci/github-branch-source-plugin/commit/ea6e28ca69ea24a59584fba3c6190b145dc0925e>
>    ** Found a way to force array initializer expansion if that is what is
>    desired - 43a9950
>    <https://github.com/jenkinsci/github-branch-source-plugin/commit/43a99508ab640db724e8b673c61855cb57670b82>
>    ** Long string arguments getting pushed to the next line? Maybe the
>    humans involved should split those up. #392 (comment)
>    <https://github.com/jenkinsci/github-branch-source-plugin/pull/392#discussion_r580403349>
>    ** I've applied the ".gitattributes" suggested by Joseph. The build
>    now passes on Windows.
>    - I have also not addressed a number of other places where people have
>    said the formatting looks "worse". These cases are generally
>    debatable/personal-preference. In some of these cases I agree with the
>    assessment but I don't see it as a blocker. If anyone disagrees with that
>    evaluation, feel free to say so and explain why.
>    ** #392 (comment)
>    <https://github.com/jenkinsci/github-branch-source-plugin/pull/392#discussion_r580104724>
>    - see
>    https://github.com/bitwiseman/github-branch-source-plugin/pull/new/task/formatting-test-cleanup
>    ** #392 (comment)
>    <https://github.com/jenkinsci/github-branch-source-plugin/pull/392#discussion_r580105388>
>    - same as previous
>    ** #392 (comment)
>    <https://github.com/jenkinsci/github-branch-source-plugin/pull/392#discussion_r580402264>
>    ** #392 (comment)
>    <https://github.com/jenkinsci/github-branch-source-plugin/pull/392#discussion_r580510020>
>    ** #392 (comment)
>    <https://github.com/jenkinsci/github-branch-source-plugin/pull/392#discussion_r580513803>
>
> Note: I am using the eclipse formatter rather than to google formatter in
> an attempt to reduce the severity of the diff wall and disruption to folks
> familiar with the current code style. There are still significant
> differences, but they are far less than switching to google formatting.
> However, that is still an option. You can see what it would look like here:
> https://github.com/bitwiseman/github-branch-source-plugin/pull/new/task/formatting-google
>
> @basil <https://github.com/basil> has made a strong case for switching to
> Google Formatter, which would obviates my tweaks above. If I follow my own
> stated reasoning to its logical conclusion, what's he's saying makes sense.
> On the other hand, the "diff wall" in that case is absolute. I need to sit
> with this for a day or two, but I will likely make that change.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jenkinsci/github-branch-source-plugin/pull/392#issuecomment-786381314>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADYZBV2CTVQZYGRBM3NQV3TA4GTPANCNFSM4X5LUTOQ>
> .
>

jglick commented:
> you just may have to look at the parent of the diff that changed the line - but that is still the case today - e.g. refactoring

Sure, this just means you would need to do that in basically every case, rather than on occasion.

I agree with @basil’s reasoning about Google Format. As to the actual formatting it produces, it seems comparable to any other machine formatting I have seen—i.e., loses human intent and is somewhat less legible, and the line length seems artificially short, but :shrug: tolerable.

(Obviously I only rarely touch this particular repo any more, so I am just kibitzing.)

bitwiseman commented:
This change will be a squash merge to keep the diff wall to one commit.
